01projects/sec02SoftwareBuilds.md

```diff
@@ -0,0 +1,86 @@
+
+The need to invoke multiple toolchains, conditionally enabling various capabilities, that are dependent on hardware introspection, for each platform and OS quickly becomes non-scalable.
+Any such executable (or in general any output artefact of the workflow) produced with specific desired properties is considered as a *software build* (or simply a *build*).
+It is evident that obtaining software builds through manual invocation of various tools either at the command-line or through basic IDE options quickly becomes cumbersome and tedious.
+
+### Build automation tools
+
+#### Makefiles
+
+The complexities of build tools have been acknowledged since nearly half a century, and several build automation tools (also known as build systems) have been devised over time to tackle the problem.
+An early such tool, which is still popular today, is [`make`](https://en.wikipedia.org/wiki/Make_(software)).
+In this build system, the library author writes a set of plain-text files, known as MakeFiles, and provides them along with the C++ source code.
+For system introspection, a set of companion tools called [GNU Autotools](https://en.wikipedia.org/wiki/GNU_Autotools) was developed to configure the build capabilities depending on the target processor's capabilities.
+
+While this works quite well (even today) for a large number of scenarios, the `Make` and `Autotools` system have a number of drawbacks by modern yardsticks.
+For instance, they are available only in Unix-like OSes. MakeFiles use various cryptic shorthand notations and are frustratingly stubborn about formatting requirements.
+The syntax makes is hard to fathom the relationships between various dependencies, and there are no built-in mechanisms to visualise them.
+There is no built-in debugger.
+In short, it is tedious to write MakeFiles by hand for large projects, but it is a good build system, as long as the required MakeFiles can somehow be generated for us.
+Enter build system generators!
+
+#### Build system generators
+
+Non-Unix build systems have their own native build systems.
+For instance on Windows, with the Microsoft C++ compiler and toolchain (MSVC), the build system is known as Visual Studio (an IDE exclusive to Windows) solutions, and on macOSes, the toolchain provided by Apple emits Xcode (an IDE exclusive to macOS) compatible output.
+
+In late 1999/early 2000, there was no compatible build system that worked across all Operating Systems, and developers had to go through the frustrating process of manually invoking multiple toolchains on different systems.
+It was during this time that, Kitware Inc, a company that was part of a consortium tasked with designing the C++-based Imaging Toolkit (itk) funded by a US national research grant, developed CMake, a cross-platform *build system generator* (a system that emits output for consumption by native build systems, which in-turn have to be invoked to produce the final build output).
```

Comment:
The word "emits" might be a bit confusing here, I would maybe use "generates" or "outputs" and then say what the output actually is e.g. make files. Otherwise I think it won't be entirely clear what CMake is actually doing to someone who has never heard of it before.

Reply:
@mmcleod89 Thank you. Have clarified this.